app/handler/unsubscribe_encoder.py

```diff
@@ -25,52 +40,106 @@ class UnsubscribeLink:
 
 class UnsubscribeEncoder:
     @staticmethod
-    def encode(action: UnsubscribeAction, data: int) -> UnsubscribeLink:
+    def encode(
+        action: UnsubscribeAction, data: Union[int, UnsubscribeOriginalData]
+    ) -> UnsubscribeLink:
         if config.UNSUBSCRIBER:
             return UnsubscribeLink(UnsubscribeEncoder.encode_mailto(action, data), True)
         return UnsubscribeLink(UnsubscribeEncoder.encode_url(action, data), False)
 
-    @staticmethod
-    def encode_subject(action: UnsubscribeAction, data: int) -> str:
-        if action == UnsubscribeAction.DisableAlias:
-            return f"{data}="
-        if action == UnsubscribeAction.DisableContact:
-            return f"{data}_"
-        if action == UnsubscribeAction.UnsubscribeNewsletter:
-            return f"{data}*"
+    @classmethod
+    def encode_subject(
+        cls, action: UnsubscribeAction, data: Union[int, UnsubscribeOriginalData]
+    ) -> str:
+        if (
+            action != UnsubscribeAction.OriginalUnsubscribeMailto
+            and type(data) is not int
+        ):
+            raise ValueError(f"Data has to be an int for an action of type {action}")
+        if action == UnsubscribeAction.OriginalUnsubscribeMailto:
+            if type(data) is not UnsubscribeOriginalData:
+                raise ValueError(
+                    f"Data has to be an UnsubscribeOriginalData for an action of type {action}"
+                )
+            # Initial 0 is the version number. If we need to add support for extra use-cases we can bump up this number
+            data = (0, data.alias_id, data.recipient, data.subject)
```

Comment:
I wonder if `version` should be part of the UnsubscribeOriginalData?

Reply:
If we need it in the future we can add it. For not it's not needed. It's a fail-safe in the encoding to be able to be backwards compatible if we need it since this info will be encoded in emails and can potentially be stored forever in user emails.